idaplugin/rematch/actions/match.py

```diff
@@ -71,60 +100,46 @@ def response_handler(self, file_version):
   def start_upload(self):
     self.functions = set(idautils.Functions())
 
-    self.pbar = QtWidgets.QProgressDialog()
     self.pbar.setLabelText("Processing IDB... You may continue working,\nbut "
                            "please avoid making any ground-breaking changes.")
     self.pbar.setRange(0, len(self.functions))
     self.pbar.setValue(0)
-    self.pbar.canceled.connect(self.cancel_upload)
-    self.pbar.rejected.connect(self.reject_upload)
     self.pbar.accepted.connect(self.accept_upload)
+    self.pbar.show()
 
-    self.timer = QtCore.QTimer()
     self.timer.timeout.connect(self.perform_upload)
     self.timer.start(0)
 
     return True
 
   def perform_upload(self):
-    try:
-      offset = self.functions.pop()
-    except KeyError:
-      self.timer.stop()
+    if not self.functions:
       return
 
-    try:
-      func = instances.FunctionInstance(self.file_version_id, offset)
-      self.instance_set.append(func.serialize())
-
-      if len(self.instance_set) >= 100:
-        network.delayed_query("POST", "collab/instances/",
-                              params=self.instance_set, json=True,
-                              callback=self.progress_advance)
-        self.instance_set = []
-        self.pbar.setMaximum(self.pbar.maximum() + 1)
-      self.progress_advance()
-    except Exception:
-      self.cancel_upload()
-      raise
+    offset = self.functions.pop()
+    func = instances.FunctionInstance(self.file_version_id, offset)
+    self.instance_set.append(func.serialize())
+
+    if len(self.instance_set) >= 100 or not self.functions:
```

Comment:
The idea here is this:
We should send objects to the server in two cases:
1. We collected 100 instances, enough to send a request for. We'd like to consolidate multiple requests into few requests with decent size.
2. If we just processed the last function (we `pop`ed the last one, and `self.functions` just became empty). We need to make sure we won't "forget" the last items if they didn't reach the bulk-size.

I know this piece of code is a bit too difficult to understand. Any ideas on how to make it clearer?

I'll also comment it so it'll be clearer.